### DISCUSSION TITLE
What is the best way to migrate existing extension settings to a new structure?

### DISCUSSION THREAD

**Discussion by sketchbuch** [microsoft/vscode-discussions#862] [Extension Development QnA] [answered]
For one of my extension the setings are growing and it is becoming a mess. I thought about structuring the config into different sections.

What is the best way to migrate existing config to a new structure?

Has anyone done this?

Is there any recommended way?

I know you can group settings, but I would also like to change keys.

**Comment by alefragnani** [comment, marked as answer]
Hi @sketchbuch ,

I remember having a good discussion about it in the Slack channel, but unfortunately, the thread isn't available anymore due to its free plan limit. If I remember correctly, there was a few caveats:

* First of all, there is no _builtin/automatic_ migration API available in VS Code. But, the `"deprecationMessage"` is your friend here
* If you can avoid migration, you should avoid 😬 
* Double check if _group settings_ isn't a good alternative for your scenario, as it does not requires any migration

But, if you really need migration, I see at least two alternatives:

### Your extension will take care of the entire migration, with no warn to the user

Sure, this could the best UX, as it is transparent to the user. But, you will have a bit more work to do, and you must be sure to not mess up with user settings. Also, I'm not sure how this could be affected by **User Profiles**, as each profile may contain different versions of the same extension

### You can create the new setting and _migrate_ the values from the old setting

Personally, this is my favorite path. Create the new setting, migrate the original value to it and use the `deprecationMessage` to warn the user about the old setting. After a few releases, simply remove the old setting.

If I remember correctly, VS Code itself used this approach too. I also did that on my extensions, and I guess GitLens used this approach in the past too.

Hope this helps

**Reply by sketchbuch** [reply]
Thanks! Yeah, I will just group my settings and leave the keys as is, this will be a lot easier. Next time I make an extension with config I think I will keep the keys as flat as possible and just group the settings.
